mock-hpos-api/index.js

```diff
@@ -0,0 +1,147 @@
+  return `${method}:${removeApiRoot(path)}:${JSON.stringify(data)}`
+}
+
+// NEXT_PATH is used to key a next response in place of an actual path. It's required to be different from any value that could come from req.path
+const NEXT_PATH = ' . next . '
+
+const NEXT_RESPONSE_KEY = generateResponseKey('', NEXT_PATH, {})
+
+class MockHposApi {
+  constructor(port, authEmail, authPassword) {
+    this.port = port
+    this.shouldCheckAuth = !!authEmail && !!authPassword
+    this.authEmail = authEmail
+    this.authPassword = authPassword
+    
+    this.initializeResponses()
+  }
+
+  static async start (port, authEmail, authPassword) {
+    const mockHposApi = new MockHposApi(port, authEmail, authPassword)
+
+    const app = express()
+
+    app.use(cors())
+    app.use(bodyParser.json())
+    app.use(mockHposApi.checkAuth.bind(mockHposApi))
+    app.use(mockHposApi.handleRequest.bind(mockHposApi))
+
+    await mockHposApi.startServer(app)
+
+    return mockHposApi
+  }
+
+  // this is more convoluted than I'd like because I want init to be a blocking call, so have to uncallback app.listen
+  startServer (app) {
+    return new Promise((resolve, reject) => {
+      this.server = app.listen(this.port, () => {
+        console.log(`Mock hpos-admin-api listening at http://localhost:${this.port}`)
+        resolve()
+      }).on('error', reject)
+    })
+  }
+
+  async checkAuth (req, res, next) {
+    if (!this.shouldCheckAuth) {
+      next()
+    } else {
+
+      const { method, path, body } = req
+
+      const keypairInstance = new HpAdminKeypair(HC_PUBKEY, this.authEmail, this.authPassword)
+
+      const signature = await signPayload(keypairInstance, method, path, body)
+
+      if (signature === req.header('x-hpos-admin-signature')) {
+        next()
+      } else {
+        res.status(401).end()
+      }
+    }
+  }
+
+  any (response) {
+    this.anyResponse = response
+  }
+
+  next (response) {
+    this.once(NEXT_PATH, {}, response)
+  }
+
+  once (path, body, response) {
+    const responseKey = generateResponseKey(method, path, body)
+
+    if (!this.responseQueues[responseKey]) {
+      this.responseQueues[responseKey] = []
+    }
+  
+    this.responseQueues[responseKey].push(response)  
+  }
+
+  getSavedResponse (method, type, data) {
+    let responseKey
+
+    // if there are responses in the 'next' queue, we use those and ignore the specific type and data of the request
+    if (!_.isEmpty(this.responseQueues[NEXT_RESPONSE_KEY])) {
+      responseKey = NEXT_RESPONSE_KEY
+    } else {
+      responseKey = generateResponseKey(method, type, data)
+    }
+
+    if (!this.responseQueues[responseKey]) {
+      if (this.anyResponse) return this.anyResponse
+
+      throw new Error(`No more responses for: ${responseKey}`)
+    }
+
+    return this.responseQueues[responseKey].shift()
+  }
+
+  initializeResponses () {
+    this.responseQueues = {}
+    this.anyResponse = defaultResponse
```

Comment:
This variable stores a function that is called whenever we don't have a queued response right? What do you think of renaming it `fallbackResponse` or `fallbackResponder`?

Reply:
😆 The variable is called `anyResponse` because it's the response set by the `.any` method. There's already been a lot of discussion about the name of the `any` method. I think `fallbackResponse` is a totally reasonable name. I'm going to leave it as `anyResponse` mostly so that it doesn't diverge from `mock-conductor` and you can re-use understanding between those repos. But if someone opened PRs that updated both I would accept it.

Reply:
would `.all` and `.allResponse` be clearer to you?

Reply:
I can imagine how this makes sense at the call site.
"The **next** request should give this response"
"**Any** other request should give that response"

It's just confusing at the definition site. No need to change it if there's been so much bikeshedding on this already.

> would `.all` and `.allResponse` be clearer to you?

No that almost seems less clear! :D

Reply:
Maybe just add comments then?

```js
// Sets the response for the next request that this mock handles

// Sets the fallback response for any requests that were not configured by a call to `next()`
```

Reply:
👍